cmd/gitserver/server/server.go

```diff
@@ -297,19 +320,57 @@ func (s *Server) SyncRepoState(interval time.Duration, batchSize, perSecond int)
 	}
 }
 
-// hostnameMatch checks whether the hostname matches the given address.
-// If we don't find an exact match, we look at the initial prefix.
-func (s *Server) hostnameMatch(addr string) bool {
-	if !strings.HasPrefix(addr, s.Hostname) {
-		return false
+// getHostname get the current hostname. It returns an error if the hostname has
+// not been set yet.
+func (s *Server) getHostname() (string, error) {
+	s.hostnameMu.RLock()
+	defer s.hostnameMu.RUnlock()
+	if s.hostname == "" {
+		return "", errors.New("hostname not set")
 	}
-	if addr == s.Hostname {
-		return true
+	return s.hostname, nil
+}
+
+// setHostnameOnce sets hostname only once if h is not blank
+// and it can be found in addrs
+func (s *Server) setHostnameOnce(h string, addrs []string) {
+	if h == "" {
+		return
+	}
+	if !hostnameFound(h, addrs) {
+		return
 	}
-	// We know that s.Hostname is shorter than addr so we can safely check the next
-	// char
-	next := addr[len(s.Hostname)]
-	return next == '.' || next == ':'
+	s.hostnameMu.Lock()
+	defer s.hostnameMu.Unlock()
+	// Another caller may have already set it
+	if s.hostname != "" {
+		return
+	}
+	s.hostname = h
+}
+
+// hostnameFound returns true only if our hostname can be found in addrs
+func hostnameFound(hostname string, addrs []string) bool {
+	for _, a := range addrs {
+		if hostname == a {
+			return true
+		}
+	}
+	return false
+}
+
+// hostnameFromFrontend returns the hostname provided from the request only if it
+// is from one of our frontend instances.
+func hostnameFromFrontend(r *http.Request) string {
```

Comment:
Tiny nitpick again, but while reading this and the other `if ... != ""` checks above I thought that maybe a second return value can help?

```suggestion
func hostnameFromFrontend(r *http.Request) string, bool {
```